tinygrad/runtime/ops_hip.py

```diff
@@ -63,22 +77,62 @@ def __init__(self, name:str, prg:bytes):
     hip.hipSetDevice(args[0]._device)
-    if wait:
-      start, end = hip.hipEventCreate(), hip.hipEventCreate()
-      hip.hipEventRecord(start)
     struct = hip.getStructTypeForArgs(*args)(*[data._buf if not isinstance(data, int) else np.int32(data) for data in args])
-    hip.hipModuleLaunchKernel(self.prgs[args[0]._device], global_size[0], global_size[1], global_size[2], local_size[0], local_size[1], local_size[2], 0, 0, struct)
-    if wait:
-      hip.hipEventRecord(end)
-      hip.hipEventSynchronize(end)
-      ret = hip.hipEventElapsedTime(start, end)*1e-3
-      hip.hipEventDestroy(start)
-      hip.hipEventDestroy(end)
-      return ret
+    return time_execution(lambda: hip.hipModuleLaunchKernel(self.prgs[args[0]._device], global_size[0], global_size[1], global_size[2], local_size[0], local_size[1], local_size[2], 0, 0, struct), enable=wait)
 
   def __del__(self):
     for module in self.modules: hip.hipModuleUnload(module)
 
+class HIPGraph(BatchExecutor):
+  UpdateNode = collections.namedtuple("UpdateNode", ["node", "params"])
+
+  def __init__(self, jit_cache: List[JitItem], input_rawbuffers: List[RawBuffer], var_vals: Dict[Variable, int]):
+    super().__init__(jit_cache, input_rawbuffers, var_vals)
+
+    # TODO: Only HIPProgram can be captured for now.
+    if not all(isinstance(ji.prg, CompiledASTRunner) and isinstance(ji.prg.clprg, HIPProgram) for ji in self.jit_cache): return
+
+    self.graph, graph_node = hip.hipGraphCreate(), None
+    self.updatable_nodes: Dict[int, HIPGraph.UpdateNode] = {}
+
+    for (j,i),input_name in self.input_replace.items(): self.jit_cache[j].rawbufs[i] = input_rawbuffers[input_name]
+    for j,ji in enumerate(self.jit_cache):
+      prg: CompiledASTRunner = cast(CompiledASTRunner, ji.prg)
+      global_size, local_size = prg.launch_dims(var_vals)
+
+      assert all(x is not None for x in ji.rawbufs) and ji.rawbufs[0] is not None, "buffers could not be None" # for linters
+      params = hip.buildKernelNodeParams(*ji.rawbufs, *[var_vals[x] for x in prg.vars], func=prg.clprg.prgs[ji.rawbufs[0]._device], grid=global_size, block=local_size)
+      graph_node = hip.hipGraphAddKernelNode(self.graph, [graph_node] if graph_node else [], params)
+
+      if j in self.jc_idxs_with_updatable_launch_dims or j in self.jc_idxs_with_updatable_var_vals or j in self.jc_idxs_with_updatable_rawbufs:
+        self.updatable_nodes[j] = HIPGraph.UpdateNode(graph_node, params)
+
+    self.instance = hip.hipGraphInstantiate(self.graph)
+    self.clear_jit_inputs()
+
+  def __del__(self):
+    if hasattr(self, 'instance'): hip.hipGraphExecDestroy(self.instance)
+    if hasattr(self, 'graph'): hip.hipGraphDestroy(self.graph)
+
+  def __call__(self, input_rawbuffers: List[RawBuffer], var_vals: Dict[Variable, int], wait=False):
+    if not hasattr(self, 'instance'): return super().__call__(input_rawbuffers, var_vals, wait)
```

Comment:
How does METAL deal with this...looks hacky

Reply:
This is a case for world.py and multigpus. We can't use graphs for jit_cache which contains memory operations and not only `HIPProgram`s. Now it just switches to basic batch executor. In the future we can encode memory operations into the graph as well and events for synchronisation.

Reply:
My concern is not that it switches, it's how this is being done with these `hasattr`